server/src/main/java/org/opensearch/indices/IndicesRequestCache.java

```diff
@@ -501,23 +523,38 @@ private void updateCleanupKeyToCountMapOnCacheEviction(CleanupKey cleanupKey) {
             }
             ShardId shardId = indexShard.shardId();
 
-            cleanupKeyToCountMap.computeIfPresent(shardId, (shard, keyCountMap) -> {
-                keyCountMap.computeIfPresent(cleanupKey.readerCacheKeyId, (key, currentValue) -> {
-                    // decrement the stale key count
-                    staleKeysCount.decrementAndGet();
-                    int newValue = currentValue - 1;
-                    // Remove the key if the new value is zero by returning null; otherwise, update with the new value.
+            cleanupKeyToCountMap.computeIfPresent(shardId, (id, stringIntegerConcurrentMap) -> {
+                stringIntegerConcurrentMap.compute(cleanupKey.readerCacheKeyId, (readerCacheKeyId, count) -> {
+                    // Check if the key is currently present
+                    if (count != null) {
+                        // The key does not exist, so this is already accounted in the staleKeysCount
+                        // and hence needs to be decremented
+                        staleKeysCount.decrementAndGet();
+                    }
+
+                    // Regardless of whether the key was initially present, we perform the decrement operation
+                    // Calculate the new value assuming a null count as zero to handle non-existent keys
+                    int newValue = (count == null ? 0 : count) - 1;
```

Comment:
When `count == null` the newValue will be -1. This doesn't seem right. We need to decrement count only when it is not null